### PR TITLE
Tests implementation

### PR DESCRIPTION
**Pull request by supermafete**

### CODE REVIEW

**Comment by jimmyn** [comment]
I think tests should be placed in the same folder as a tested file. This makes them easier to find.

Check `src/lib`

**Comment by jimmyn** [comment]
yes, this is the convention

File: jest.config.js [hunk 1]
```diff
@@ -2,12 +2,19 @@
...
 module.exports = {
-  moduleFileExtensions: ['ts', 'js'],
-  testMatch: ['**/*.test.(ts|js)'],
-  moduleDirectories: ['node_modules'],
+  roots: [
+    '<rootDir>/src'
+  ],
   transform: {
-    '^.+\\.(js|ts)$': 'babel-jest'
+    '^.+\\.tsx?$': 'ts-jest'
```
Comment (jimmyn) [inline jest.config.js]: you don't need to change anything there, babel already uses babel-typescript and compiles everything as needed.

**Comment by supermafete** [comment]
@rhinoandre take a look

File: package.json [hunk 2]
```diff
@@ -49,7 +50,8 @@
     "joi": "^14.0.4",
     "joi-error-formatter": "^0.5.3",
     "middy": "^0.19.2",
-    "source-map-support": "^0.5.9"
+    "source-map-support": "^0.5.9",
+    "ts-jest": "^23.10.4"
```
Comment (jimmyn) [inline package.json]: why do you need `ts-jest`? `@babel/preset-typescript` works quite well, with source maps and all.

File: src/users/listUsers/listUsers.test.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  userModelloadAll.mockReturnValue(userModel);
+  userModelexecAsync.mockReturnValue(userModel);
+
+   return {
+     UserModel: userModel
+   }
+});
+
+describe('UserModel', () => {
+  it('Check if all functions are executed correctly', () => {
```
Comment (jimmyn) [inline src/users/listUsers/listUsers.test.ts]: What you need to test here is that the response form `listUsers` is correct when you mock dynamoDB request. The internal implementation could change, but the result should be the same in order for client to work correctly. You can use [matchSnapshot()](https://jestjs.io/docs/en/snapshot-testing)  jest method.

File: tsconfig.json [hunk 4]
```diff
@@ -21,6 +21,7 @@
   },
   "include": [
     "src/**/*",
+    "tests/**/*",
```
Comment (jimmyn) [inline tsconfig.json]: you don't need this

File: src/lib/validation.test.ts [hunk 5]
```diff
@@ -3,6 +3,8 @@ import {assert, attempt, getEnv, Joi} from './validation';
 // define env variables for your test at the top
 process.env.AWS_REGION = 'eu-west-1';
 
+console.log(process.env.USERS_TABLE);
```
Comment (jimmyn) [inline src/lib/validation.test.ts]: you don't need this

File: src/users/listUsers/handler.ts [hunk 6]
```diff
@@ -3,6 +3,8 @@ import { apiRequestRoutine } from '../../lib/middlewares/apiRequestRoutine';
 import { logRoutine } from '../../lib/middlewares/logRoutine';
 import { listUsers } from './listUsers';
 
+process.env.AWS_REGION = 'eu-west-1';
```
Comment (jimmyn) [inline src/users/listUsers/handler.ts]: you don't need this

File: jest.config.js [hunk 7]
```diff
@@ -6,7 +6,7 @@ module.exports = {
   testMatch: ['**/*.test.(ts|js)'],
   moduleDirectories: ['node_modules'],
   transform: {
-    '^.+\\.(js|ts)$': 'babel-jest'
+    '^.+\\.(js|ts)$': 'ts-jest'
```
Comment (jimmyn) [inline jest.config.js]: please revert this so I can merge it

**Comment by jimmyn** [comment]
Please merge with master and resolve conflicts

File: src/users/listUsers/listUsers.test.ts (hunk 3 above)
Comment (jimmyn) [reply src/users/listUsers/listUsers.test.ts]: @rhinoandre agreed

**Comment by jimmyn** [comment]
lets follow a naming convention

```js
describe('functionName', () => {
   it('should do something', () => {
      return expect(functionName()).resolves.toMatchSnapshot();
   }
})
```

And you can actually use expect(...).resolves, but remember to return it, as it is also a promise

File: jest.config.js (hunk 1 above)
Comment (jimmyn) [reply jest.config.js]: not resolved

File: package.json (hunk 2 above)
Comment (jimmyn) [reply package.json]: not resolved

File: tsconfig.json (hunk 4 above)
Comment (jimmyn) [reply tsconfig.json]: not resolved

File: src/lib/validation.test.ts (hunk 5 above)
Comment (jimmyn) [reply src/lib/validation.test.ts]: not resolved

File: src/users/listUsers/handler.ts (hunk 6 above)
Comment (jimmyn) [reply src/users/listUsers/handler.ts]: not resolved

**Comment by supermafete** [comment]
sorry, wrong resolve. try now.

**Pull request merged**